rust/kernel/module_param.rs

```diff
@@ -0,0 +1,137 @@
+use core::fmt::Write;
+
+/// Types that can be used for module parameters.
+/// Note that displaying the type in `sysfs` will fail if `to_string` returns
+/// more than 4K bytes (including an additional null terminator).
+pub trait ModuleParam : core::fmt::Display + core::marker::Sized {
+    fn try_from_param_arg(arg: &[u8]) -> Option<Self>;
+
+    /// # Safety
+    ///
+    /// `val` must point to a valid null-terminated string. The `arg` field of
+    /// `param` must be an instance of `Self`.
+    unsafe extern "C" fn set_param(val: *const crate::c_types::c_char, param: *const crate::bindings::kernel_param) -> crate::c_types::c_int {
+        let arg = crate::c_types::c_string_bytes(val);
+        match Self::try_from_param_arg(arg) {
+            Some(new_value) => {
+                let old_value = (*param).__bindgen_anon_1.arg as *mut Self;
+                let _ = core::ptr::replace(old_value, new_value);
+                0
+            }
+            None => crate::error::Error::EINVAL.to_kernel_errno()
+        }
+    }
+
+    /// # Safety
+    ///
+    /// `buf` must be a buffer of length at least `kernel::PAGE_SIZE` that is
+    /// writeable. The `arg` field of `param` must be an instance of `Self`.
+    unsafe extern "C" fn get_param(buf: *mut crate::c_types::c_char, param: *const crate::bindings::kernel_param) -> crate::c_types::c_int {
+        let slice = core::slice::from_raw_parts_mut(buf as *mut u8, crate::PAGE_SIZE);
+        let mut buf = crate::buffer::Buffer::new(slice);
+        match write!(buf, "{}\0", *((*param).__bindgen_anon_1.arg as *mut Self)) {
+            Err(_) => crate::error::Error::EINVAL.to_kernel_errno(),
+            Ok(()) => buf.bytes_written() as crate::c_types::c_int,
+        }
+    }
+
+    /// # Safety
+    ///
+    /// The `arg` field of `param` must be an instance of `Self`.
+    unsafe extern "C" fn free(arg: *mut crate::c_types::c_void) {
+        core::ptr::drop_in_place(arg as *mut Self);
+    }
+}
+
+/// Trait for parsing integers. Strings begining with `0x`, `0o`, or `0b` are
```

Comment:
I assume you went for a trait vs. a generic function to make it look like `from_str_radix` for callers, right?

Reply:
I think we need to make a trait anyway. There's no trait (at least not in `std`) for types that implement `from_str_radix` or `checked_neg` so if we wanted to make a generic function we'd have to create that trait so we could use it as a constraint on the argument for that function. Seems to me that either way would be reasonable.

Reply:
Right, there is `num` out there, but it seems like Rust could use a `PrimitiveInteger` trait or similar.